File: packages/tonic-core/src/hydration/buildIntentHydration.ts
import * as fs from "node:fs";
import * as path from "node:path";

import type { AstHydrationArtifactV1, NormalizedMatch } from "../astGrep/types";
import type { ConflictContextArtifactV1 } from "./conflictScan";
import type { IntentBootstrapArtifactV1 } from "./intentBootstrap";
import type { QuestionRefinementArtifactV1 } from "./questionRefinement";

export type RetrievalHitBrief = {
  chunk_id: string;
  text?: string;
  score?: number;
  metadata?: Record<string, unknown>;
};

export type IntentHydrationArtifactV1 = {
  schema: "tonic-intent-hydration";
  version: "1";
  left_intent: string;
  right_intent: string;
  truncation_policy_version?: string;
  token_budget_hint?: number;
  conflict_region_ids?: string[];
  evidence_links: Array<{
    id?: string;
    type: "ast_match" | "retrieval_hit" | "code_walk" | "conflict";
    path: string;
    line_start?: number;
    line_end?: number;
    detail?: string;
  }>;
  prompt_excerpt?: string;
};

function pickIntents(
  boot: IntentBootstrapArtifactV1,
  refine: QuestionRefinementArtifactV1 | null,
): { left: string; right: string } {
  if (
    refine &&
    refine.mode !== "off" &&
    refine.refined_left_intent &&
    refine.refined_right_intent
  ) {
    return { left: refine.refined_left_intent, right: refine.refined_right_intent };
  }
  if (refine?.refined_left_intent && refine.refined_right_intent) {
    return { left: refine.refined_left_intent, right: refine.refined_right_intent };
  }
  return { left: boot.left_intent, right: boot.right_intent };
}

function astMatchesNearConflicts(
  ast: AstHydrationArtifactV1 | null,
  conflicts: ConflictContextArtifactV1,
  maxPerFile: number,
): NormalizedMatch[] {
  if (!ast || ast.matches.length === 0 || conflicts.conflict_regions.length === 0) {
    return ast?.matches.slice(0, 50) ?? [];
  }
  const byFile = new Map<string, ConflictContextArtifactV1["conflict_regions"]>();
  for (const r of conflicts.conflict_regions) {
    const list = byFile.get(r.path) ?? [];
    list.push(r);
    byFile.set(r.path, list);
  }
  const scored: Array<{ m: NormalizedMatch; score: number }> = [];
  for (const m of ast.matches) {
    const regs = byFile.get(m.path);
    if (!regs) {
      continue;
    }
    const line = m.start?.line ?? 0;
    let best = Infinity;
    for (const reg of regs) {
      const d = Math.abs(line - reg.mid_line);
      if (d < best) {
        best = d;
      }
    }
    scored.push({ m, score: best });
  }
  scored.sort((a, b) =>
    a.score !== b.score
      ? a.score - b.score
      : a.m.path !== b.m.path
        ? a.m.path.localeCompare(b.m.path)
        : (a.m.rule_id || "").localeCompare(b.m.rule_id || ""),
  );
  const out: NormalizedMatch[] = [];
  const perFile = new Map<string, number>();
  for (const { m } of scored) {
    const n = (perFile.get(m.path) ?? 0) + 1;
    if (n > maxPerFile) {
      continue;
    }
    perFile.set(m.path, n);
    out.push(m);
    if (out.length >= 80) {
      break;
    }
  }
  if (out.length === 0) {
    return ast.matches.slice(0, 50);
  }
  return out;
}

export function buildIntentHydration(params: {
  bootstrap: IntentBootstrapArtifactV1;
  refinement: QuestionRefinementArtifactV1 | null;
  conflicts: ConflictContextArtifactV1 | null;
  ast: AstHydrationArtifactV1 | null;
  astPath?: string;
  conflictPath?: string;
  retrieval?: { artifactPath: string; hits: RetrievalHitBrief[] };
  codeWalkTracePath?: string;
}): IntentHydrationArtifactV1 {
  const { left, right } = pickIntents(params.bootstrap, params.refinement);
  const links: IntentHydrationArtifactV1["evidence_links"] = [];
  const regionIds: string[] = [];

  if (params.conflicts) {
    for (const r of params.conflicts.conflict_regions) {
      if (r.region_id) {
        regionIds.push(r.region_id);
      }
      links.push({
        id: r.region_id,
        type: "conflict",
        path: r.path,
        line_start: r.start_line,
        line_end: r.end_line,
        detail: "git conflict marker region",
      });
    }
  }

  const emptyConflicts: ConflictContextArtifactV1 = {
    schema: "tonic-conflict-context",
    version: "1",
    scan_scope: "none",
    conflict_regions: [],
  };
  const near = astMatchesNearConflicts(params.ast, params.conflicts ?? emptyConflicts, 5);
  let i = 0;
  for (const m of near) {
    i += 1;
    links.push({
      id: `ast-${i}`,
      type: "ast_match",
      path: m.path,
      line_start: m.start?.line,
      line_end: m.end?.line,
      detail: m.rule_id,
    });
  }

  const maxRetrievalLinks = 25;
  if (params.retrieval?.hits?.length) {
    for (let j = 0; j < Math.min(params.retrieval.hits.length, maxRetrievalLinks); j++) {
      const h = params.retrieval.hits[j]!;
      const p = String(h.metadata?.path ?? "");
      const meta = h.metadata ?? {};
      const boost = meta.ast_boost_applied === true ? " ast_boost_applied=true" : "";
      const near = typeof meta.nearest_conflict_region_id === "string" ? meta.nearest_conflict_region_id : "";
      const nearPart = near ? ` nearest_conflict_region_id=${near}` : "";
      links.push({
        id: `retrieval-${h.chunk_id}`,
        type: "retrieval_hit",
        path: p,
        line_start: typeof h.metadata?.start_line === "number" ? h.metadata.start_line : undefined,
        line_end: typeof h.metadata?.end_line === "number" ? h.metadata.end_line : undefined,
        detail: `score=${typeof h.score === "number" ? h.score.toFixed(4) : "?"}${boost}${nearPart}`,
      });
    }
  }

  if (params.codeWalkTracePath) {
    links.push({
      id: "code-walk-trace",
      type: "code_walk",
      path: params.codeWalkTracePath.replace(/\\/g, "/"),
      detail: "tonic-code-walk-trace.v1",
    });
  }

  const excerpt =
    `Intents — left: ${left}\n` +
    `Intents — right: ${right}\n` +
    (params.refinement?.subquestions?.length
      ? `Subquestions: ${params.refinement.subquestions.map((s) => s.text).join(" | ")}\n`
      : "") +
    `Evidence links: ${links.length}`;

  links.sort((a, b) => {
    const ta = a.type.localeCompare(b.type);
    if (ta !== 0) {
      return ta;
    }
    const pa = a.path.localeCompare(b.path);
    if (pa !== 0) {
      return pa;
    }
    return String(a.id ?? "").localeCompare(String(b.id ?? ""));
  });

  return {
    schema: "tonic-intent-hydration",
    version: "1",
    left_intent: left,
    right_intent: right,
    truncation_policy_version: "1",
    token_budget_hint: 8000,
    conflict_region_ids: regionIds.length ? regionIds : undefined,
    evidence_links: links,
    prompt_excerpt: excerpt,
  };
}

export function readJsonFile<T>(p: string): T | null {
  try {
    return JSON.parse(fs.readFileSync(p, "utf8")) as T;
  } catch {
    return null;
  }
}

export function writeIntentHydration(pathOut: string, art: IntentHydrationArtifactV1): void {
  fs.mkdirSync(path.dirname(path.resolve(pathOut)), { recursive: true });
  fs.writeFileSync(pathOut, JSON.stringify(art, null, 2) + "\n", "utf8");
}

File: merge-tonic-lib/tonic/hydration/build_intent_hydration.py
"""Merge prior artifacts into tonic-intent-hydration.v1 (parity with buildIntentHydration.ts)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from tonic.hydration.sensitive_artifact_redaction import redact_sensitive_values


def _pick_intents(boot: dict[str, Any], refine: dict[str, Any] | None) -> tuple[str, str]:
    if refine:
        mode = refine.get("mode")
        rl = refine.get("refined_left_intent")
        rr = refine.get("refined_right_intent")
        if mode != "off" and isinstance(rl, str) and isinstance(rr, str) and rl and rr:
            return rl, rr
        if isinstance(rl, str) and isinstance(rr, str) and rl and rr:
            return rl, rr
    return str(boot.get("left_intent", "")), str(boot.get("right_intent", ""))


def _ast_matches_near_conflicts(
    ast: dict[str, Any] | None, conflicts: dict[str, Any], max_per_file: int
) -> list[dict[str, Any]]:
    matches = list(ast.get("matches") or []) if ast else []
    regs = list(conflicts.get("conflict_regions") or [])
    if not matches or not regs:
        return matches[:50]

    by_file: dict[str, list[dict[str, Any]]] = {}
    for r in regs:
        if not isinstance(r, dict):
            continue
        p = str(r.get("path", ""))
        by_file.setdefault(p, []).append(r)

    scored: list[tuple[dict[str, Any], float]] = []
    for m in matches:
        if not isinstance(m, dict):
            continue
        path = str(m.get("path", ""))
        if path not in by_file:
            continue
        start = m.get("start") or {}
        line = int(start.get("line", 0)) if isinstance(start, dict) else 0
        best = float("inf")
        for reg in by_file[path]:
            mid = int(reg.get("mid_line", 0))
            d = abs(line - mid)
            if d < best:
                best = d
        scored.append((m, best))

    scored.sort(
        key=lambda x: (
            x[1],
            str(x[0].get("path", "")),
            str(x[0].get("rule_id", "")),
        )
    )
    out: list[dict[str, Any]] = []
    per_file: dict[str, int] = {}
    for m, _ in scored:
        path = str(m.get("path", ""))
        n = per_file.get(path, 0) + 1
        if n > max_per_file:
            continue
        per_file[path] = n
        out.append(m)
        if len(out) >= 80:
            break
    return out if out else matches[:50]


def build_intent_hydration(
    *,
    bootstrap: dict[str, Any],
    refinement: dict[str, Any] | None,
    conflicts: dict[str, Any] | None,
    ast: dict[str, Any] | None,
    retrieval: dict[str, Any] | None = None,
    code_walk_trace_path: str | None = None,
) -> dict[str, Any]:
    left, right = _pick_intents(bootstrap, refinement)
    links: list[dict[str, Any]] = []
    region_ids: list[str] = []

    cc = conflicts or {"conflict_regions": []}
    for r in cc.get("conflict_regions") or []:
        if not isinstance(r, dict):
            continue
        rid = r.get("region_id")
        if rid:
            region_ids.append(str(rid))
        links.append(
            {
                "id": rid,
                "type": "conflict",
                "path": str(r.get("path", "")),
                "line_start": r.get("start_line"),
                "line_end": r.get("end_line"),
                "detail": "git conflict marker region",
            }
        )

    empty_cc = {"conflict_regions": []}
    near = _ast_matches_near_conflicts(ast, conflicts or empty_cc, 5)
    for i, m in enumerate(near, start=1):
        start = m.get("start") or {}
        end = m.get("end") or {}
        ls = start.get("line") if isinstance(start, dict) else None
        le = end.get("line") if isinstance(end, dict) else None
        links.append(
            {
                "id": f"ast-{i}",
                "type": "ast_match",
                "path": str(m.get("path", "")),
                "line_start": ls,
                "line_end": le,
                "detail": str(m.get("rule_id", "")),
            }
        )

    max_retrieval_links = 25
    if retrieval and isinstance(retrieval.get("hits"), list):
        for j, h in enumerate(retrieval["hits"][:max_retrieval_links]):
            if not isinstance(h, dict):
                continue
            meta = h.get("metadata") or {}
            p = str(meta.get("path", ""))
            ls = meta.get("start_line")
            le = meta.get("end_line")
            sc = h.get("score")
            detail = f"score={float(sc):.4f}" if isinstance(sc, (int, float)) else "score=?"
            if meta.get("ast_boost_applied") is True:
                detail += " ast_boost_applied=true"
            near = meta.get("nearest_conflict_region_id")
            if isinstance(near, str) and near.strip():
                detail += f" nearest_conflict_region_id={near.strip()}"
            links.append(
                {
                    "id": f"retrieval-{h.get('chunk_id', j)}",
                    "type": "retrieval_hit",
                    "path": p,
                    "line_start": int(ls) if isinstance(ls, int) else None,
                    "line_end": int(le) if isinstance(le, int) else None,
                    "detail": detail,
                }
            )

    if code_walk_trace_path:
        links.append(
            {
                "id": "code-walk-trace",
                "type": "code_walk",
                "path": code_walk_trace_path.replace("\\", "/"),
                "detail": "tonic-code-walk-trace.v1",
            }
        )

    sub_excerpt = ""
    if refinement and isinstance(refinement.get("subquestions"), list):
        sq = refinement["subquestions"]
        texts = [str(x.get("text", "")) for x in sq if isinstance(x, dict)]
        if texts:
            sub_excerpt = "Subquestions: " + " | ".join(texts) + "\n"

    excerpt = (
        f"Intents — left: {left}\n"
        f"Intents — right: {right}\n"
        f"{sub_excerpt}"
        f"Evidence links: {len(links)}"
    )

    links.sort(
        key=lambda x: (
            str(x.get("type", "")),
            str(x.get("path", "")),
            str(x.get("id", "") or ""),
        )
    )

    out: dict[str, Any] = {
        "schema": "tonic-intent-hydration",
        "version": "1",
        "left_intent": left,
        "right_intent": right,
        "truncation_policy_version": "1",
        "token_budget_hint": 8000,
        "evidence_links": links,
        "prompt_excerpt": excerpt,
    }
    if region_ids:
        out["conflict_region_ids"] = region_ids
    return out


def write_intent_hydration(path_out: str, art: dict[str, Any]) -> None:
    p = Path(path_out).resolve()
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(redact_sensitive_values(art), indent=2) + "\n", encoding="utf-8")
